### PR TITLE
Add filtering to Sessions().ListAll()

### PR DESCRIPTION
**Pull request by petergeorgas**
## Type of change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 🌟 New feature (non-breaking change which adds functionality)
- [ ] 🔨 Breaking change (fix or feature that would cause existing functionality)
- [ ] 📖 Docs change / refactoring / dependency upgrade to change)

## Description

Add Filtering to `Sessions().ListAll()` to match the API. 

### Related Issue (optional)

#125

### CODE REVIEW

File: clerk/sessions.go [hunk 1]
```diff
@@ -15,10 +15,47 @@ type Session struct {
...
+	STATUS_ABANDONED SessionStatus = "abandoned"
+	STATUS_ACTIVE    SessionStatus = "active"
+	STATUS_ENDED     SessionStatus = "ended"
+	STATUS_EXPIRED   SessionStatus = "expired"
+	STATUS_REMOVED   SessionStatus = "removed"
+	STATUS_REPLACED  SessionStatus = "replaced"
+	STATUS_REVOKED   SessionStatus = "revoked"
+)
+
+func (s *SessionsService) ListAll(params ListAllSessionsParams) ([]Session, error) {
```
Comment (dimkl) [inline clerk/sessions.go:38]: 🔧 i would expect the params to be optional to avoid making this a BREAKING CHANGE.
```
ListAll(params *ListAllSessionsParams) 
```

File: clerk/sessions_test.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ func TestSessionsService_ListAll_happyPath(t *testing.T) {
 	var want []Session
 	_ = json.Unmarshal([]byte(expectedResponse), &want)
 
-	got, _ := client.Sessions().ListAll()
+	got, _ := client.Sessions().ListAll(ListAllSessionsParams{})
```
Comment (dimkl) [inline clerk/sessions_test.go:26]: 🔧 revert these changes to test that current sdk usage does not break.

File: clerk/sessions_test.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ import (
 	"testing"
 )
 
-func TestSessionsService_ListAll_happyPath(t *testing.T) {
+func TestSessionsService_ListAll_happyPath_noParams(t *testing.T) {
```
Comment (dimkl) [inline clerk/sessions_test.go:11]: 🔧 Could you please add 2 tests:
- `TestSessionsService_ListAll_pagination_and_filtering_params`
- `TestSessionsService_ListAll_pagination_and_filtering_empty_params`

test to assert that the added functionality works as expected?

File: clerk/sessions.go (hunk 1 above)
Comment (gkats) [reply clerk/sessions.go:38]: As @dimkl mentioned, simply adding the params here would be a breaking change. We're not ready for a major version bump yet unfortunately.

Even adding them as a pointer would be a breaking change, so we can't do `ListAll(p *ListAllSessionsParams)`.

I think the correct way to approach this would be to add a separate method on the SessionsService. Something like

```go
func (s *SessionService) ListAllWithFiltering(params ListAllWithFilteringSessionsParams) {}
```

@petergeorgas What do you think about that? Open to any other suggestion of course. The important thing is to be backwards compatible.

Comment (petergeorgas) [reply clerk/sessions.go:38]: Good point! Just made the changes!

File: clerk/sessions.go [hunk 4]
```diff
@@ -15,6 +15,26 @@ type Session struct {
...
+
+const (
+	STATUS_ABANDONED SessionStatus = "abandoned"
+	STATUS_ACTIVE    SessionStatus = "active"
+	STATUS_ENDED     SessionStatus = "ended"
+	STATUS_EXPIRED   SessionStatus = "expired"
+	STATUS_REMOVED   SessionStatus = "removed"
+	STATUS_REPLACED  SessionStatus = "replaced"
+	STATUS_REVOKED   SessionStatus = "revoked"
+)
```
Comment (gkats) [inline clerk/sessions.go:36]: :wrench: Consider prefixing these constants with `SESSION_`. Since everything's inside the Clerk package, if we ever want to add similar constants to other resources, there's going to be a naming collision.

Also, let's write them in camel case, maybe it's more Go idiomatic.

```go
// This is for sessions
const SessionStatusAbandoned SessionStatus = "abandoned"

// This is for some other resource
const OtherResourceStatusAbandoned = "abandoned"

// They are both exported by the clerk package
clerk.OtherResourceStatusAbandoned
clerk.SessionStatusAbandoned
```

**Review by gkats — CHANGES_REQUESTED** [review]
Awesome work @petergeorgas!

Left one more comment regarding the session status constants. Let's update this and then we're good to :shipit: 

Thank you so much for taking care of this!

File: clerk/sessions.go (hunk 4 above)
Comment (petergeorgas) [reply clerk/sessions.go:36]: Sounds good! I never know what to do with constants... thanks!

**Review by gkats — APPROVED** [review]

**Comment by gkats** [comment]
@petergeorgas Thank you so much for the contribution!! :heart_eyes: :tada:

**Comment by petergeorgas** [comment]
> @petergeorgas Thank you so much for the contribution!! 😍 🎉

Happy to help! Looking forward to tackling other issues in the future.
